service/raft/raft_sys_table_storage.cc

```diff
@@ -306,9 +305,20 @@ future<> raft_sys_table_storage::abort() {
     return std::move(_pending_op_fut);
 }
 
-future<> raft_sys_table_storage::truncate_log_tail(raft::index_t idx) {
-    static const auto truncate_cql = format("DELETE FROM system.{} WHERE group_id = ? AND \"index\" <= ?", db::system_keyspace::RAFT);
-    return _qp.execute_internal(truncate_cql, {_group_id.id, int64_t(idx)}, cql3::query_processor::cache_internal::yes).discard_result();
+future<> raft_sys_table_storage::update_snapshot_and_truncate_log_tail(const raft::snapshot_descriptor &snap, size_t preserve_log_entries) {
+    // Update snapshot and truncate logs in `system.raft` atomically
+    raft::index_t log_tail_idx = raft::index_t(static_cast<uint64_t>(snap.idx) - static_cast<uint64_t>(preserve_log_entries));
+    static const auto store_latest_id_and_truncate_log_tail_cql = format(
+        "BEGIN UNLOGGED BATCH"
+        "   INSERT INTO system.{} (group_id, snapshot_id) VALUES (?, ?);"   // store latest id
+        "   DELETE FROM system.{} WHERE group_id = ? AND \"index\" <= ?;"   // truncate log tail
+        "APPLY BATCH",
+        db::system_keyspace::RAFT, db::system_keyspace::RAFT);
```

Comment:
Maybe it's not necessary, but I think it would be more elegant to remove the `preserve_log_entries > snap.idx`  case, and keep the logic of storing snapshot_id only inside `update_snapshot_and_truncate_log_tail`.

Then you could do it by always constructing a batch statement; but in the case of `preserve_log_entries > snap.idx`, it would be a batch consisting of only the single statement to store snapshot_id. Under the condition `preserve_log_entries <= snap.idx` you would add the truncating statement to the batch.

WDYT?

I leave the decision to you, I think it's mostly a matter of style.

Reply:
> remove the preserve_log_entries > snap.idx case

from `store_snapshot_descriptor`

Reply:
I took some time and tried the other way. But it looks like I'm not quite there yet with cpp to wrangle those statement value pointers.